source/extensions/filters/http/jwt_authn/filter.h

```diff
@@ -0,0 +1,52 @@
+
+#pragma once
+
+#include "envoy/http/filter.h"
+
+#include "common/common/logger.h"
+
+#include "extensions/filters/http/jwt_authn/authenticator.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace JwtAuthn {
+
+// The Envoy filter to process JWT auth.
+class Filter : public Http::StreamDecoderFilter,
+               public Authenticator::Callbacks,
+               public Logger::Loggable<Logger::Id::filter> {
+public:
+  Filter(DataStoreFactorySharedPtr store_factory);
+
+  // Http::StreamFilterBase
+  void onDestroy() override;
+
+  // Http::StreamDecoderFilter
+  Http::FilterHeadersStatus decodeHeaders(Http::HeaderMap& headers, bool) override;
+  Http::FilterDataStatus decodeData(Buffer::Instance&, bool) override;
+  Http::FilterTrailersStatus decodeTrailers(Http::HeaderMap&) override;
+  void setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) override;
+
+private:
+  // the function for Authenticator::Callbacks interface.
+  // It will be called when its verify() call is completed.
+  void onComplete(const ::google::jwt_verify::Status& status) override;
+
+  // The callback funcion.
+  Http::StreamDecoderFilterCallbacks* decoder_callbacks_;
+  // store_factory object
+  DataStoreFactorySharedPtr store_factory_;
+  // The auth object.
+  AuthenticatorPtr auth_;
+  // The state of the request
+  enum State { Init, Calling, Responded, Complete };
```

Comment:
@saumoh this is similar to what you do in ext_authz. Is there a bug here or is this a simpler way of looking at the side-call state?